### PR TITLE
Search for matches between a color image patch and an input color image

### PR DESCRIPTION
**Pull request by LaurentBerger**
**Merge with extra**: https://github.com/opencv/opencv_extra/pull/534

Sangwine, S. J., Ell, T. A. and Moxey C. E., Vector Phase Correlation, Electronics Letters, 37(25), 6 December 2001, 1513-5. http://dx.doi.org/10.1049/el:20011035

T. A. Ell and S. J. Sangwine. Decomposition of 2D hypercomplex Fourier transforms into pairs of complex Fourier transforms. In Moncef Gabbouj and Pauli Kuosmanen, editors, Proceedings of EUSIPCO 2000, Tenth European Signal Processing Conference, volume II, pages 1061–1064, Tampere, Finland, 5–8 September 2000. European Association for Signal Processing

Quaternion Toolbox for Matlab r . [Online]. Software library, written by Stephen Sangwine and Nicolas Le Bihan. Available: http://qtfm.sourceforge.net/.

### CODE REVIEW

**Comment by LaurentBerger** [comment]
PR Not Ready

**Comment by alalek** [comment]
@LaurentBerger Is this PR ready?

**Comment by LaurentBerger** [comment]
@alalek no I will try to write some test next week

**Comment by alalek** [comment]
Try to squash commits in opencv_extra (to drop intermediate history at least).

File: modules/ximgproc/include/opencv2/ximgproc/color_match.hpp [hunk 1]
```diff
@@ -0,0 +1,68 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#ifndef __OPENCV_COLOR_MATCH_HPP__
+#define __OPENCV_COLOR_MATCH_HPP__
+#ifdef __cplusplus
```
Comment (alalek) [inline modules/ximgproc/include/opencv2/ximgproc/color_match.hpp]: `#ifdef __cplusplus` is not needed in .hpp files.

File: modules/ximgproc/samples/color_match_template.cpp [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    }
+    flip(dst, dst, -1);
+
+    imshow("Max Quaternion corr",dst);
+}
+
+int main(int , char *[])
+{
+    Mat imgLogo = imread("g:/lib/opencv/samples/data/opencv-logo.png", IMREAD_COLOR);
+    Mat imgColor = imread("g:/lib/opencv/samples/data/lena.jpg", IMREAD_COLOR);
```
Comment (alalek) [inline modules/ximgproc/samples/color_match_template.cpp]: Please avoid using of absolute paths. `samples/data/...` should be enough (or just file names).

File: modules/ximgproc/src/quaternion.cpp [hunk 3]
```diff
@@ -0,0 +1,206 @@
+﻿#include "precomp.hpp"
```
Comment (alalek) [inline modules/ximgproc/src/quaternion.cpp]: Please add license header (modules sources / headers / tests, but not samples):

```
// This file is part of OpenCV project.
// It is subject to the license terms in the LICENSE file found in the top-level directory
// of this distribution and at http://opencv.org/license.html.
```

File: modules/ximgproc/src/quaternion.cpp [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+using namespace std;
+using namespace cv;
```
Comment (alalek) [inline modules/ximgproc/src/quaternion.cpp]: `using namespace cv;` is not needed. All code is placed into `cv` namespace.

File: modules/ximgproc/src/quaternion.cpp [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+
+namespace cv
+{
+namespace ximgproc {
+
+void createQuaternionImage(InputArray _img, OutputArray _qimg)
+{
+    int type = _img.type(), depth = CV_MAT_DEPTH(type), cn = CV_MAT_CN(type);
+    CV_Assert((depth == CV_8U || depth == CV_32F || depth == CV_64F) && _img.dims() == 2 && cn == 3);
```
Comment (alalek) [inline modules/ximgproc/src/quaternion.cpp]: It is better to use `CV_Check*()` macros for parameters checks.

File: modules/ximgproc/src/quaternion.cpp [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+void qconj(InputArray _img, OutputArray _qimg)
+{
+    int type = _img.type(), depth = CV_MAT_DEPTH(type), cn = CV_MAT_CN(type);
+    CV_Assert((depth == CV_32F || depth == CV_64F) && _img.dims() == 2 && cn == 4);
+    vector<Mat> qplane(4), plane;
+    split(_img, plane);
+    qplane[0] = plane[0].clone();
+    qplane[1] = -plane[1].clone();
+    qplane[2] = -plane[2].clone();
+    qplane[3] = -plane[3].clone();
```
Comment (alalek) [inline modules/ximgproc/src/quaternion.cpp]: Perhaps `.clone()` is not necessary here.

File: modules/ximgproc/test/test_matchcolortemplate.cpp [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    Mat img, logo;
+    img = imread(openCVExtraDir + "cv/ximgproc/image.png", IMREAD_COLOR);
+    logo = imread(openCVExtraDir + "cv/ximgproc/opencv_logo.png", IMREAD_COLOR);
+    ximgproc::colorMatchTemplate(img, logo, corr);
+
+    absdiff(corr, corrRef, dd);
+    double maxVal;
+    Point pIdx;
+    minMaxLoc(dd, NULL, &maxVal, NULL, &pIdx);
+    ASSERT_LE(maxVal, 1e-5);
```
Comment (alalek) [inline modules/ximgproc/test/test_matchcolortemplate.cpp]: 5 lines -> `EXPECT_LE(cv::norm(corr, corrRef, NORM_INF), 1e-5);`

File: modules/ximgproc/src/quaternion.cpp [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+    qplane[1] = plane[1].clone();
+    qplane[2] = plane[2].clone();
+    qplane[3] = plane[3].clone();
+    double *ptr0 = qplane[0].ptr<double>(0, 0), *ptr1 = qplane[1].ptr<double>(0, 0);
+    double *ptr2 = qplane[2].ptr<double>(0, 0), *ptr3 = qplane[3].ptr<double>(0, 0);
+    int nb = plane[0].rows*plane[0].cols;
+    for (int i = 0; i < nb; i++, ptr0++, ptr1++, ptr2++, ptr3++)
+    {
+        double d = *ptr0 * *ptr0 + *ptr1 * *ptr1 + *ptr2 * *ptr2 + *ptr3 * *ptr3;
+        d = 1 / sqrt(d);
```
Comment (alalek) [inline modules/ximgproc/src/quaternion.cpp]: Try parallel `.forEach()`:
```
    _img.copyTo(_qimg);
    Mat qimg = _qimg.getMat();
    qimg.forEach<Vec4d>([](Vec4d &p, const int * /*position*/) -> void {
        double d = p[0] * p[0] + p[1] * p[1] + p[2] * p[2] + p[3] * p[3];
        d = 1 / sqrt(d);
        p *= d;
    });
```

File: modules/ximgproc/src/quaternion.cpp [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+        *ptr1 *= d;
+        *ptr2 *= d;
+        *ptr3 *= d;
+    }
+    merge(qplane, _qimg);
+}
+
+void qdft(InputArray _img, OutputArray _qimg, int  	flags, bool sideLeft)
+{
+    //    CV_INSTRUMENT_REGION()
```
Comment (alalek) [inline modules/ximgproc/src/quaternion.cpp]: It is OK to add this (into all exported "large enough" functions).
With ';' in the end.

File: modules/ximgproc/include/opencv2/ximgproc/color_match.hpp (hunk 1 above)
Comment (LaurentBerger) [reply modules/ximgproc/include/opencv2/ximgproc/color_match.hpp]: ok

File: modules/ximgproc/src/quaternion.cpp (hunk 3 above)
Comment (LaurentBerger) [reply modules/ximgproc/src/quaternion.cpp]: OK

File: modules/ximgproc/src/quaternion.cpp (hunk 4 above)
Comment (LaurentBerger) [reply modules/ximgproc/src/quaternion.cpp]: OK

File: modules/ximgproc/src/quaternion.cpp (hunk 8 above)
Comment (LaurentBerger) [reply modules/ximgproc/src/quaternion.cpp]: Ok

File: modules/ximgproc/src/quaternion.cpp (hunk 6 above)
Comment (LaurentBerger) [reply modules/ximgproc/src/quaternion.cpp]: OK

File: modules/ximgproc/test/test_matchcolortemplate.cpp (hunk 7 above)
Comment (LaurentBerger) [reply modules/ximgproc/test/test_matchcolortemplate.cpp]: OK

File: modules/ximgproc/src/quaternion.cpp (hunk 8 above)
Comment (LaurentBerger) [reply modules/ximgproc/src/quaternion.cpp]: OK

File: modules/ximgproc/include/opencv2/ximgproc/color_match.hpp (hunk 1 above)
Comment (LaurentBerger) [reply modules/ximgproc/include/opencv2/ximgproc/color_match.hpp]: OK

File: modules/ximgproc/src/quaternion.cpp (hunk 3 above)
Comment (LaurentBerger) [reply modules/ximgproc/src/quaternion.cpp]: Ok

File: modules/ximgproc/src/quaternion.cpp (hunk 4 above)
Comment (LaurentBerger) [reply modules/ximgproc/src/quaternion.cpp]: OK

**Comment by LaurentBerger** [comment]
@alalek  can you help me for http://pullrequest.opencv.org/buildbot/builders/precommit-contrib_docs/builds/12103

File: modules/ximgproc/include/opencv2/ximgproc/color_match.hpp [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+//! @addtogroup ximgproc_filters
+//! @{
+
+    /**
+    * @brief   creates a quaternion image.
+    *
+    * @param   img         Source 8-bit, 32-bit or 64-bit image, with 3-channel image.
+    * @param   qimg        result CV_64FC4 a quaternion image( 4 chanels zero channel and B,G,R).
+    */
+    CV_EXPORTS_W void createQuaternionImage(InputArray img, OutputArray qimg);
```
Comment (alalek) [inline modules/ximgproc/include/opencv2/ximgproc/color_match.hpp]: Please avoid extra indentation in namespaces

File: modules/ximgproc/src/quaternion.cpp [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+    _img.copyTo(_qimg);
+    Mat qimg = _qimg.getMat();
+    qimg.forEach<Vec4d>([](Vec4d &p, const int * /*position*/) -> void {
+        double d = p[0] * p[0] + p[1] * p[1] + p[2] * p[2] + p[3] * p[3];
+        d = 1 / sqrt(d);
+        p *= d;
+    });
+    /*
+    qplane[0] = plane[0].clone();
+    qplane[1] = plane[1].clone();
```
Comment (alalek) [inline modules/ximgproc/src/quaternion.cpp]: I believe we can remove commented out code.

File: modules/ximgproc/src/quaternion.cpp [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+    qplane[3] = -plane[3];
+    merge(qplane, _qimg);
+}
+
+void qunitary(InputArray _img, OutputArray _qimg)
+{
+    int type = _img.type(), depth = CV_MAT_DEPTH(type), cn = CV_MAT_CN(type);
+    CV_Assert((depth == CV_64F) && _img.dims() == 2 && cn == 4);
+    vector<Mat> qplane(4), plane;
+    split(_img, plane);
```
Comment (alalek) [inline modules/ximgproc/src/quaternion.cpp]: `qplane` and `plane` are not needed anymore

**Pull request merged**